consensus/src/lib.rs

```diff
@@ -0,0 +1,96 @@
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use simperby_common::{
+    crypto::{Hash256, PublicKey},
+    BlockHeight, ConsensusRound, Timestamp,
+};
+use simperby_network::*;
+use std::collections::{HashMap, HashSet};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("network error: {0}")]
+    Network(simperby_network::Error),
+    #[error("unknown error: {0}")]
+    Unknown(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ConsensusState {
+    pub non_nil_votes: HashMap<Hash256, HashSet<PublicKey>>,
+    pub nil_votes: HashMap<Hash256, HashSet<PublicKey>>,
+    pub height: BlockHeight,
+    pub round: ConsensusRound,
+}
+
+pub enum ProgressResult {
+    Proposed(ConsensusRound, Hash256, Timestamp),
+    NonNilPreVoted(ConsensusRound, Hash256, Timestamp),
+    NonNilPreComitted(ConsensusRound, Hash256, Timestamp),
+    NilPreVoted(ConsensusRound, Timestamp),
+    NilPreComitted(ConsensusRound, Timestamp),
+    Finalized(Timestamp),
```

Comment:
How about making a variant like `NothingToProgress` instead of returning an error from `progress`?

Reply:
replaced with `Vec`

Reply:
cool